packages/react-utilities/src/compose/types.ts

```diff
@@ -48,3 +49,31 @@ export type SlotProps<TSlots extends BaseSlots, TProps, TRootProps extends React
 } & {
   root: TRootProps;
 };
+
+/**
+ * Helper type to convert the given props of type ShorthandProps into ObjectShorthandProps
+ */
+export type ResolvedShorthandProps<T, K extends keyof T> = Omit<T, K> &
+  { [P in K]: T[P] extends ShorthandProps<infer U> ? ObjectShorthandProps<U> : T[P] };
+
+/**
+ * Helper type to mark the given props as required.
+ * Similar to Required<T> except it only requires a subset of the props.
+ */
+export type RequiredProps<T, K extends keyof T> = Omit<T, K> & { [P in K]-?: T[P] };
+
+/**
+ * Converts a components Props type to a State type:
+ * * Adds the 'ref' and 'as' props
+ * * Ensures the specified ShorthandProps are of type ObjectShorthandProps<T>
+ * * Marks the given DefaultedProps as required (-?)
+ */
+export type ComponentState<
+  Props,
+  ShorthandProps extends keyof Props = never,
+  DefaultedProps extends keyof ResolvedShorthandProps<Props, ShorthandProps> = never,
+  RefType = React.Ref<HTMLElement>
```

Comment:
Current default value satisfies both `ref`s (from `React.forwardRef()`) and merged refs (`ref: useMergedRefs()`), IMO it can be confusing:

```tsx
const state = mergeProps({ ref: useMergedRefs() })
state.ref // will be React.Ref... 
```

Do I correctly understand?

---

```suggestion
  RefType = never
```
```suggestion
  RefType = unknown
```

Can we use `never` or `unknown` so user will need to specify a proper type if it is a case?

Reply:
This is a continuation of the conversation on the RFC (https://github.com/microsoft/fluentui/pull/17232#discussion_r590604969). My assumption is that most(?) components will simply forward the ref, and don't need to call `useMergedRefs`.

The previous default (from the create-component template) of `React.RefObject<HTMLElement>`  _requires_ you to use `useMergedRefs`, even if you never need it. That seems like an unnecessary perf hit.

If we wanted to force the dev to specify the ref type in the template, it'd probably best to not have any default value rather than defaulting to `never`. That said, I still think having `state.ref` be simply a forwarded `React.Ref` seems to be the most common and useful case, which is why I had it as a default. But I can change it to not have a default if that seems like a better option?

Reply:
I am voting for `never` as default value 🙂

Reply:
Could you explain why `never` as a default is better than not having any default? If the goal is to always make the dev specify the type, then having no default seems like the best option because it will give an error at an expected place (missing template parameter), rather than a confusing error about `ref` being the wrong type or something.

Reply:
No default also works and will be better for DX 👍

Reply:
I second this, probably the right thing to do (also, I'd prefer going for `never` since it's in line with the ones above).

Reply:
Ok I've updated it so that RefType is the first type argument, and doesn't have any default value.